<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using pRoloc for spatial proteomics data analysis • pRoloc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using pRoloc for spatial proteomics data analysis">
<meta property="og:description" content="pRoloc">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pRoloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.31.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-pRoloc-tutorial.html">Using pRoloc for spatial proteomics data analysis</a>
    </li>
    <li>
      <a href="../articles/v02-pRoloc-ml.html">Machine learning techniques available in pRoloc</a>
    </li>
    <li>
      <a href="../articles/v03-pRoloc-bayesian.html">Bayesian Analysis of Spatial Proteomics data using pRoloc</a>
    </li>
    <li>
      <a href="../articles/v04-pRoloc-goannotations.html">Annotating spatial proteomics data</a>
    </li>
    <li>
      <a href="../articles/v05-pRoloc-transfer-learning.html">A transfer learning algorithm for spatial proteomics</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/pRoloc/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v01-pRoloc-tutorial_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using pRoloc for spatial proteomics data analysis</h1>
                        <h4 data-toc-skip class="author">Lisa M. Breckels</h4>
            <address class="author_afil">
      Computational Proteomics Unit, Cambridge, UK<br><h4 data-toc-skip class="author">Laurent Gatto</h4>
            <address class="author_afil">
      de Duve Institute, UCLouvain, Belgium<br><small class="dont-index">Source: <a href="https://github.com/lgatto/pRoloc/blob/master/vignettes/v01-pRoloc-tutorial.Rmd" class="external-link"><code>vignettes/v01-pRoloc-tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>v01-pRoloc-tutorial.Rmd</code></div>

    </address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This tutorial illustrates the usage of the <code>pRoloc</code> R package for the analysis and interpretation of spatial proteomics data. It walks the reader through the creation of <em>MSnSet</em> instances, that hold the quantitative proteomics data and meta-data and introduces several aspects of data analysis, including data visualisation and application of machine learning to predict protein localisation.</p>
    </div>
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="foreword" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#foreword" class="anchor" aria-hidden="true"></a>Foreword</h1>
<p><em><a href="http://bioconductor.org/packages/MSnbase" class="external-link">MSnbase</a></em> and <em><a href="http://bioconductor.org/packages/pRoloc" class="external-link">pRoloc</a></em> are under active developed; current functionality is evolving and new features are added on a regular basis. This software is free and open-source software. If you use it, please support the project by citing it in publications:</p>
<blockquote>
<p>Gatto L. and Lilley K.S. <em>MSnbase - an R/Bioconductor package for isobaric tagged mass spectrometry data visualization, processing and quantitation</em>. <a href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btr645" class="external-link">Bioinformatics 28, 288-289 (2011)</a>.</p>
</blockquote>
<blockquote>
<p>Gatto L, Breckels LM, Wieczorek S, Burger T, Lilley KS. <em>Mass-spectrometry-based spatial proteomics data analysis using pRoloc and pRolocdata.</em> <a href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btu013" class="external-link">Bioinformatics. 2014 May 1;30(9):1322-4.</a>.</p>
</blockquote>
<blockquote>
<p>Breckels LM, Mulvey CM, Lilley KS and Gatto L. A Bioconductor workflow for processing and analysing spatial proteomics data. F1000Research 2016, 5:2926 <a href="https://f1000research.com/articles/5-2926/" class="external-link">doi: 10.12688/f1000research.10411.1</a>.</p>
</blockquote>
<p>If you are using the <em>phenoDisco</em> function, please also cite</p>
<blockquote>
<p>Breckels L.M., Gatto L., Christoforou A., Groen A.J., Kathryn Lilley K.S. and Trotter M.W. <em>The effect of organelle discovery upon sub-cellular protein localisation.</em> J Proteomics, <a href="http://www.sciencedirect.com/science/article/pii/S1874391913000948" class="external-link">S1874-3919(13)00094-8 (2013)</a>.</p>
</blockquote>
<p>If you are using any of the transfer learning functions, please also cite</p>
<blockquote>
<p>Breckels LM, Holden S, Wonjar D, Mulvey CM, Christoforou A, Groen A, Trotter MW, Kohlbacker O, Lilley KS and Gatto L (2016). <em>Learning from heterogeneous data sources: an application in spatial proteomics</em>. PLoS Comput Biol 13;12(5):e1004920. doi: <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004920" class="external-link">10.1371/journal.pcbi.1004920</a>.</p>
</blockquote>
<p>If you are using any of the Bayesian generative models, please also cite</p>
<blockquote>
<p>A Bayesian Mixture Modelling Approach For Spatial Proteomics Oliver M Crook, Claire M Mulvey, Paul D. W. Kirk, Kathryn S Lilley, Laurent Gatto bioRxiv 282269; doi: <a href="https://doi.org/10.1101/282269" class="external-link uri">https://doi.org/10.1101/282269</a></p>
</blockquote>
<p>For an introduction to spatial proteomics data analysis:</p>
<blockquote>
<p>Gatto L, Breckels LM, Burger T, Nightingale DJ, Groen AJ, Campbell C, Nikolovski N, Mulvey CM, Christoforou A, Ferro M, Lilley KS. <em>A foundation for reliable spatial proteomics data analysis</em>. Mol Cell Proteomics. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4125728/" class="external-link">2014 Aug;13(8):1937-52</a>. doi: 10.1074/mcp.M113.036350.</p>
</blockquote>
<p>The <em><a href="http://bioconductor.org/packages/pRoloc" class="external-link">pRoloc</a></em> package contains additional vignettes and reference material:</p>
<ul>
<li>
<em>pRoloc-tutorial</em>: pRoloc tutorial (this vignette).</li>
<li>
<em>pRoloc-ml</em>: Machine learning techniques available in pRoloc.</li>
<li>
<em>pRoloc-transfer-learning</em>: A transfer learning algorithm for spatial proteomics.</li>
<li>
<em>pRoloc-goannotations</em>: Annotating spatial proteomics data.</li>
<li>
<em>pRoloc-bayesian</em>: Bayesian spatial proteomics with pRoloc.</li>
</ul>
</div>
<div id="questions-and-bugs" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#questions-and-bugs" class="anchor" aria-hidden="true"></a>Questions and bugs</h1>
<p>You are welcome to contact me directly about <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em>. For bugs, typos, suggestions or other questions, please file an issue in our issue tracking system (<a href="https://github.com/lgatto/pRoloc/issues" class="external-link uri">https://github.com/lgatto/pRoloc/issues</a>) providing as much information as possible, a reproducible example and the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code>.</p>
<p>If you wish to reach a broader audience for general questions about proteomics analysis using <code>R</code>, you may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="external-link uri">https://support.bioconductor.org/</a>.</p>
</div>
<div id="sec:intro" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:intro" class="anchor" aria-hidden="true"></a>Introduction</h1>
<div id="spatial-proteomics" class="section level2">
<h2 class="hasAnchor">
<a href="#spatial-proteomics" class="anchor" aria-hidden="true"></a>Spatial proteomics</h2>
<p>Spatial (or organelle) proteomics is the study of the localisation of proteins inside cells. The sub-cellular compartment can be organelles, i.e. structures defined by lipid bi-layers,macro-molecular assemblies of proteins and nucleic acids or large protein complexes. In this document, we will focus on mass-spectrometry based approaches that assay a population of cells, as opposed as microscopy based techniques that monitor single cells, as the former is the primary concern of <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em>, although the techniques described below and the infrastructure in place could also be applied the processed image data. The typical experimental use-case for using <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> is a set of fractions, originating from a total cell lysate. These fractions can originate from a continuous gradient, like in the LOPIT <span class="citation">(Dunkley et al. 2006)</span> or PCP <span class="citation">(Foster et al. 2006)</span> approaches, or can be discrete fractions. The content of the fractions is then identified and quantified (using labelled or un-labelled quantitation techniques). Using relative quantitation of known organelle residents, termed organelle markers, organelle-specific profiles along the gradient are determined and new residents are identified based on matching of these distribution profiles. See for example <span class="citation">(Gatto et al. 2010)</span> and references therein for a detailed review on organelle proteomics.</p>
<p>It should be noted that large protein complexes, that are not necessarily separately enclosed within their own lipid bi-layer, can be detected by such techniques, as long as a distinct profile can be defined across the fractions.</p>
</div>
<div id="about-r-and-proloc" class="section level2">
<h2 class="hasAnchor">
<a href="#about-r-and-proloc" class="anchor" aria-hidden="true"></a>About R and <em>pRoloc</em>
</h2>
<p>R <span class="citation">(R Development Core Team 2011)</span> is a statistical programming language and interactive working environment. It can be expanded by so-called packages to confer new functionality to users. Many such packages have been developed for the analysis of high-throughput biology, notably through the Bioconductor project <span class="citation">(Gentleman et al. 2004)</span>. Two packages are of particular interest here, namely <em><a href="https://bioconductor.org/packages/3.13/MSnbase" class="external-link">MSnbase</a></em> <span class="citation">(Gatto and Lilley 2012)</span> and <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em>. The former provides flexible infrastructure to store and manipulate quantitative proteomics data and the associated meta-data and the latter implements specific algorithmic technologies to analyse organelle proteomics data.</p>
<p>Among the advantages of R are robust statistical procedures, good visualisation capabilities, excellent documentation, reproducible research<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, power and flexibility of the R language and environment itself and a rich environment for specialised functionality in many domains of bioinformatics: tools for many omics technologies, including proteomics, bio-statistics, gene ontology and biological pathway analysis, … Although there exists some specific graphical user interfaces (GUI), interaction with R is executed through a command line interface. While this mode of interaction might look alien to new users, experience has proven that after a first steep learning curve, great results can be achieved by non-programmers. Furthermore, specific and general documentation is plenty and beginners and advanced course material are also widely available.</p>
<p>Once R is started, the first step to enable functionality of a specific packages is to load them using the <code>library</code> function, as shown in the code chunk below:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://lgatto.github.io/MSnbase" class="external-link">"MSnbase"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/pRoloc" class="external-link">"pRoloc"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/pRolocdata" class="external-link">"pRolocdata"</a></span><span class="op">)</span></code></pre></div>
<p><em><a href="https://bioconductor.org/packages/3.13/MSnbase" class="external-link">MSnbase</a></em> implements the data containers that are used by <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em>. <em><a href="https://bioconductor.org/packages/3.13/pRolocdata" class="external-link">pRolocdata</a></em> is a data package that supplies several published organelle proteomics data sets.</p>
<p>As a final setup step, we set the default colour palette for some of our custom plotting functionality to use semi-transparent colours in the code chunk below (see <em>?setStockcol</em> for details). This facilitates visualisation of overlapping points.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getStockcol.html">setStockcol</a></span><span class="op">(</span><span class="cn">NULL</span><span class="op">)</span> <span class="co">## reset first</span>
<span class="fu"><a href="../reference/getStockcol.html">setStockcol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="../reference/getStockcol.html">getStockcol</a></span><span class="op">(</span><span class="op">)</span>, <span class="fl">70</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="sec:data" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:data" class="anchor" aria-hidden="true"></a>Data structures</h1>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor" aria-hidden="true"></a>Example data</h2>
<p>The data used in this tutorial has been published in <span class="citation">(Tan et al. 2009)</span>. The LOPIT technique <span class="citation">(Dunkley et al. 2006)</span> is used to localise integral and associated membrane proteins in <em>Drosophila melanogaster</em> embryos. Briefly, embryos were collected at 0 – 16 hours, homogenised and centrifuged to collect the supernatant, removing cell debris and nuclei. Membrane fractionation was performed on a iodixanol gradient and fractions were quantified using iTRAQ isobaric tags <span class="citation">(Ross et al. 2004)</span> as follows: fractions 4/5, 114; fractions 12/13, 115; fraction 19, 116 and fraction 21, 117. Labelled peptides were then separated using cation exchange chromatography and analysed by LS-MS/MS on a QSTAR XL quadrupole-time-of-flight mass spectrometer (Applied Biosystems). The original localisation analysis was performed using partial least square discriminant analysis (PLS-DA). Relative quantitation data was retrieved from the supplementary file <code>pr800866n_si_004.xls</code> (<a href="http://pubs.acs.org/doi/suppl/10.1021/pr800866n/suppl_file/pr800866n_si_004.xls" class="external-link uri">http://pubs.acs.org/doi/suppl/10.1021/pr800866n/suppl_file/pr800866n_si_004.xls</a>) and imported into R as described below. We will concentrate on the first replicate.</p>
</div>
<div id="importing-and-loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#importing-and-loading-data" class="anchor" aria-hidden="true"></a>Importing and loading data</h2>
<p>This section illustrates how to import data in comma-separated value (csv) format into an appropriate R data structure. The first section shows the original <code>csv</code> (comma separated values) spreadsheet, as published by the authors, and how one can read such a file into  using the <em>read.csv</em> function. This spreadsheet file is similar to the output of many quantitation software.</p>
<p>In the next section, we show 2 <code>csv</code> files containing a subset of the columns of original <code>pr800866n_si_004-rep1.csv</code> file and another short file, created manually, that will be used to create the appropriate R data.</p>
<div id="sec:orgcsv" class="section level3">
<h3 class="hasAnchor">
<a href="#sec:orgcsv" class="anchor" aria-hidden="true"></a>The original data file</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## The original data for replicate 1, available</span>
<span class="co">## from the pRolocdata package</span>
<span class="va">f0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>,
      pattern <span class="op">=</span> <span class="st">"pr800866n_si_004-rep1.csv"</span><span class="op">)</span>
<span class="va">csv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f0</span><span class="op">)</span></code></pre></div>
<p>The three first lines of the original spreadsheet, containing the data for replicate one, are illustrated below (using the function <em>head</em>). It contains 888 rows (proteins) and 16 columns, including protein identifiers, database accession numbers, gene symbols, reporter ion quantitation values, information related to protein identification, …</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">csv</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##   Protein.ID        FBgn Flybase.Symbol No..peptide.IDs Mascot.score
## 1    CG10060 FBgn0001104    G-ialpha65A               3       179.86
## 2    CG10067 FBgn0000044         Act57B               5       222.40
## 3    CG10077 FBgn0035720        CG10077               5       219.65
##   No..peptides.quantified area.114 area.115 area.116 area.117
## 1                       1 0.379000 0.281000 0.225000 0.114000
## 2                       9 0.420000 0.209667 0.206111 0.163889
## 3                       3 0.187333 0.167333 0.169667 0.476000
##   PLS.DA.classification Peptide.sequence Precursor.ion.mass
## 1                    PM                                    
## 2                    PM                                    
## 3                                                          
##   Precursor.ion.charge pd.2013 pd.markers
## 1                           PM    unknown
## 2                           PM    unknown
## 3                      unknown    unknown</code></pre>
</div>
<div id="sec:csv" class="section level3">
<h3 class="hasAnchor">
<a href="#sec:csv" class="anchor" aria-hidden="true"></a>From <code>csv</code> files to R data</h3>
<p>There are several ways to create the desired R data object, termed <em>MSnSet</em>, that will be used to perform the actual sub-cellular localisation prediction. Here, we illustrate a method that uses separate spreadsheet files for quantitation data, feature meta-data and sample (fraction) meta-data and the <code>readMSnSet</code> constructor function, that will hopefully be the most straightforward for new users.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## The quantitation data, from the original data</span>
<span class="va">f1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"exprsFile.csv"</span><span class="op">)</span>
<span class="va">exprsCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f1</span><span class="op">)</span>
<span class="co">## Feature meta-data, from the original data</span>
<span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"fdataFile.csv"</span><span class="op">)</span>
<span class="va">fdataCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f2</span><span class="op">)</span>
<span class="co">## Sample meta-data, a new file</span>
<span class="va">f3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRolocdata"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"pdataFile.csv"</span><span class="op">)</span>
<span class="va">pdataCsv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="va">f3</span><span class="op">)</span></code></pre></div>
<ul>
<li>
<code>exprsFile.csv</code> containing the quantitation (expression) data for the 888 proteins and 4 reporter tags.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">exprsCsv</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##          FBgn     X114     X115     X116     X117
## 1 FBgn0001104 0.379000 0.281000 0.225000 0.114000
## 2 FBgn0000044 0.420000 0.209667 0.206111 0.163889
## 3 FBgn0035720 0.187333 0.167333 0.169667 0.476000</code></pre>
<ul>
<li>
<code>fdataFile.csv</code> containing meta-data for the 888 features (here proteins).</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fdataCsv</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##          FBgn ProteinID FlybaseSymbol NoPeptideIDs MascotScore
## 1 FBgn0001104   CG10060   G-ialpha65A            3      179.86
## 2 FBgn0000044   CG10067        Act57B            5      222.40
## 3 FBgn0035720   CG10077       CG10077            5      219.65
##   NoPeptidesQuantified PLSDA
## 1                    1    PM
## 2                    9    PM
## 3                    3</code></pre>
<ul>
<li>
<code>pdataFile.csv</code> containing samples (here fractions) meta-data. This simple file has been created manually.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pdataCsv</span></code></pre></div>
<pre><code>##   sampleNames Fractions
## 1        X114       4/5
## 2        X115     12/13
## 3        X116        19
## 4        X117        21</code></pre>
<p>A self-contained data structure, called <em>MSnSet</em> (defined in the <em><a href="https://bioconductor.org/packages/3.13/MSnbase" class="external-link">MSnbase</a></em> package) can now easily be generated using the <em>readMSnSet</em> constructor, providing the respective <code>csv</code> file names shown above and specifying that the data is comma-separated (with <code>sep = ","</code>). Below, we call that object <code>tan2009r1</code> and display its content.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tan2009r1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/readMSnSet.html" class="external-link">readMSnSet</a></span><span class="op">(</span>exprsFile <span class="op">=</span> <span class="va">f1</span>,
            featureDataFile <span class="op">=</span> <span class="va">f2</span>,
            phenoDataFile <span class="op">=</span> <span class="va">f3</span>,
            sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span>
<span class="va">tan2009r1</span></code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X114 X115 X116 X117
##   varLabels: Fractions
##   varMetadata: labelDescription
## featureData
##   featureNames: FBgn0001104 FBgn0000044 ... FBgn0001215 (888 total)
##   fvarLabels: ProteinID FlybaseSymbol ... PLSDA (6 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## - - - Processing information - - -
## Quantitation data loaded: Sat May  1 09:28:30 2021  using readMSnSet. 
##  MSnbase version: 2.17.7</code></pre>
</div>
</div>
<div id="a-shorter-input-work-flow" class="section level2">
<h2 class="hasAnchor">
<a href="#a-shorter-input-work-flow" class="anchor" aria-hidden="true"></a>A shorter input work flow</h2>
<p>The <code>readMSnSet2</code> function provides a simplified import pipeline. It takes a single spreadsheet as input (default is <code>csv</code>) and extract the columns identified by <code>ecol</code> to create the expression data, while the others are used as feature meta-data. <code>ecol</code> can be a <code>character</code> with the respective column labels or a numeric with their indices. In the former case, it is important to make sure that the names match exactly. Special characters like <code>'-'</code> or <code>'('</code> will be transformed by R into <code>'.'</code> when the <code>csv</code> file is read in. Optionally, one can also specify a column to be used as feature names. Note that these must be unique to guarantee the final object validity.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ecol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"area"</span>, <span class="fl">114</span><span class="op">:</span><span class="fl">117</span>, sep <span class="op">=</span> <span class="st">"."</span><span class="op">)</span>
<span class="va">fname</span> <span class="op">&lt;-</span> <span class="st">"Protein.ID"</span>
<span class="va">eset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/readMSnSet.html" class="external-link">readMSnSet2</a></span><span class="op">(</span><span class="va">f0</span>, <span class="va">ecol</span>, <span class="va">fname</span><span class="op">)</span>
<span class="va">eset</span></code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData: none
## featureData
##   featureNames: CG10060 CG10067 ... CG9983 (888 total)
##   fvarLabels: Protein.ID FBgn ... pd.markers (12 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.17.7</code></pre>
<p>The <code>ecol</code> columns can also be queried interactively from R using the <code>getEcols</code> and <code>grepEcols</code> function. The former return a character with all column names, given a splitting character, i.e. the separation value of the spreadsheet (typically <code>","</code> for <code>csv</code>, <code>"\t"</code> for <code>tsv</code>, …). The latter can be used to grep a pattern of interest to obtain the relevant column indices.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/grepEcols.html" class="external-link">getEcols</a></span><span class="op">(</span><span class="va">f0</span>, <span class="st">","</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "\"Protein ID\""              "\"FBgn\""                   
##  [3] "\"Flybase Symbol\""          "\"No. peptide IDs\""        
##  [5] "\"Mascot score\""            "\"No. peptides quantified\""
##  [7] "\"area 114\""                "\"area 115\""               
##  [9] "\"area 116\""                "\"area 117\""               
## [11] "\"PLS-DA classification\""   "\"Peptide sequence\""       
## [13] "\"Precursor ion mass\""      "\"Precursor ion charge\""   
## [15] "\"pd.2013\""                 "\"pd.markers\""</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/grepEcols.html" class="external-link">grepEcols</a></span><span class="op">(</span><span class="va">f0</span>, <span class="st">"area"</span>, <span class="st">","</span><span class="op">)</span></code></pre></div>
<pre><code>## [1]  7  8  9 10</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/grepEcols.html" class="external-link">grepEcols</a></span><span class="op">(</span><span class="va">f0</span>, <span class="st">"area"</span>, <span class="st">","</span><span class="op">)</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/readMSnSet.html" class="external-link">readMSnSet2</a></span><span class="op">(</span><span class="va">f0</span>, <span class="va">e</span><span class="op">)</span></code></pre></div>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData: none
## featureData
##   featureNames: 1 2 ... 888 (888 total)
##   fvarLabels: Protein.ID FBgn ... pd.markers (12 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
## Annotation:  
## - - - Processing information - - -
##  MSnbase version: 2.17.7</code></pre>
<p>The <code>phenoData</code> slot can now be updated accordingly using the replacement functions <code>phenoData&lt;-</code> or <code>pData&lt;-</code> (see <code><a href="https://lgatto.github.io/MSnbase/reference/MSnSet-class.html" class="external-link">?MSnSet</a></code> for details).</p>
<div id="the-msnset-class" class="section level3">
<h3 class="hasAnchor">
<a href="#the-msnset-class" class="anchor" aria-hidden="true"></a>The <em>MSnSet</em> class</h3>
<p>Although there are additional specific sub-containers for additional meta-data (for instance to make the object MIAPE compliant), the feature (the sub-container, or slot <code>featureData</code>) and sample (the <code>phenoData</code> slot) are the most important ones. They need to meet the following validity requirements (see <a href="#fig:msnset">figure below</a>):</p>
<ul>
<li><p>the number of row in the expression/quantitation data and feature data must be equal and the row names must match exactly, and</p></li>
<li><p>the number of columns in the expression/quantitation data and number of row in the sample meta-data must be equal and the column/row names must match exactly.</p></li>
</ul>
<p>It is common, in the context of <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> to update the feature meta-data (described in section @ref(sec:analysis)) by adding new columns, without breaking the objects validity. Similarly, the sample meta-data can also be updated by adding new sample variables. A detailed description of the <code>MSnSet</code> class is available by typing <code><a href="https://lgatto.github.io/MSnbase/reference/MSnSet-class.html" class="external-link">?MSnSet</a></code> in the R console.</p>
<div class="figure">
<img src="Figures/msnset.png" id="fig:msnset" alt=""><p class="caption">Dimension requirements for the respective expression, feature and sample meta-data slots.</p>
</div>
<p>The individual parts of this data object can be accessed with their respective accessor methods:</p>
<ul>
<li>the quantitation data can be retrieved with <code>exprs(tan2009r1)</code>,</li>
<li>the feature meta-data with <code>fData(tan2009r1)</code> and</li>
<li>the sample meta-data with <code>pData(tan2009r1)</code>.</li>
</ul>
<p>The advantage of this structure is that it can be manipulated as a whole and the respective parts of the data object will remain compatible. The code chunk below, for example, shows how to extract the first 5 proteins and 2 first samples:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">smallTan</span> <span class="op">&lt;-</span> <span class="va">tan2009r1</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/pSet-class.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">smallTan</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 5 2</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">smallTan</span><span class="op">)</span></code></pre></div>
<pre><code>##                 X114     X115
## FBgn0001104 0.379000 0.281000
## FBgn0000044 0.420000 0.209667
## FBgn0035720 0.187333 0.167333
## FBgn0003731 0.247500 0.253000
## FBgn0029506 0.216000 0.183000</code></pre>
<p>Several data sets, including the 3 replicates from <span class="citation">(Tan et al. 2009)</span>, are distributed as <em>MSnSet</em> instances in the <em><a href="https://bioconductor.org/packages/3.13/pRolocdata" class="external-link">pRolocdata</a></em> package. Others include, among others, the <em>Arabidopsis thaliana</em> LOPIT data from <span class="citation">(Dunkley et al. 2006)</span> (<code>dunkley2006</code>) and the mouse PCP data from <span class="citation">(Foster et al. 2006)</span> (<code>foster2006</code>). Each data set can be loaded with the <code>data</code> function, as show below for the first replicate from <span class="citation">(Tan et al. 2009)</span>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span></code></pre></div>
<p>The original marker proteins are available as a feature meta-data variables called <code>markers.orig</code> and the output of the partial least square discriminant analysis, applied in the original publication, in the <code>PLSDA</code> feature variable. The most up-to-date marker list for the experiment can be found in <code>markers</code>. This feature meta-data column can be added from a simple <code>csv</code> markers files using the <code>addMarkers</code> function - see <code><a href="../reference/addMarkers.html">?addMarkers</a></code> for details.</p>
<p>The organelle markers are illustrated below using the convenience function <em>getMarkers</em>, but could also be done manually by accessing feature variables directly using <em>fData()</em>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getMarkers.html">getMarkers</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers.orig"</span><span class="op">)</span></code></pre></div>
<pre><code>## organelleMarkers
##            ER         Golgi mitochondrion            PM       unknown 
##            20             6            14            15           833</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getMarkers.html">getMarkers</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"PLSDA"</span><span class="op">)</span></code></pre></div>
<pre><code>## organelleMarkers
##      ER/Golgi mitochondrion            PM       unknown 
##           235            74           180           399</code></pre>
<p><strong>Important</strong> As can be seen above, some proteins are labelled <code>"unknown"</code>, defining non marker proteins. This is a convention in many <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> functions. Missing annotations (empty string) will not be considered as of unknown localisation; we prefer to avoid empty strings and make the absence of known localisation explicit by using the <code>"unknown"</code> tag. This information will be used to separate marker and non-marker (unlabelled) proteins when proceeding with data visualisation and clustering (sections @ref(sec:viz) and @ref(sec:usml)) and classification analysis (section @ref(sec:sml)).</p>
</div>
</div>
<div id="prolocs-organelle-markers" class="section level2">
<h2 class="hasAnchor">
<a href="#prolocs-organelle-markers" class="anchor" aria-hidden="true"></a><em>pRoloc</em>’s organelle markers</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> package distributes a set of markers that have been obtained by mining the <em><a href="https://bioconductor.org/packages/3.13/pRolocdata" class="external-link">pRolocdata</a></em> datasets and curation by various members of the <a href="http://proteomics.bio.cam.ac.uk/" class="external-link">Cambridge Centre for Proteomics</a>. The available marker sets can be obtained and loaded using the <code>pRolocmarkers</code> function:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pRolocmarkers.html">pRolocmarkers</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## 7 marker lists available:
## Arabidopsis thaliana [atha]:
##  Ids: TAIR, 543 markers
## Drosophila melanogaster [dmel]:
##  Ids: Uniprot, 179 markers
## Gallus gallus [ggal]:
##  Ids: IPI, 102 markers
## Homo sapiens [hsap]:
##  Ids: Uniprot, 872 markers
## Mus musculus [mmus]:
##  Ids: Uniprot, 937 markers
## Saccharomyces cerevisiae [scer_sgd]:
##  Ids: SGD, 259 markers
## Saccharomyces cerevisiae [scer_uniprot]:
##  Ids: Uniprot Accession, 259 markers</code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/pRolocmarkers.html">pRolocmarkers</a></span><span class="op">(</span><span class="st">"dmel"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##       Q7JZN0       Q7KLV9       Q9VIU7       P15348       Q7KMP8       O01367 
##         "ER" "Proteasome"         "ER"    "Nucleus" "Proteasome"    "Nucleus"</code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="../reference/pRolocmarkers.html">pRolocmarkers</a></span><span class="op">(</span><span class="st">"dmel"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##  Cytoskeleton            ER         Golgi      Lysosome mitochondrion 
##             7            24             7             8            15 
##       Nucleus    Peroxisome            PM    Proteasome  Ribosome 40S 
##            21             4            25            14            22 
##  Ribosome 60S 
##            32</code></pre>
<p>These markers can then be added to a new <em>MSnSet</em> using the <em>addMarkers</em> function by matching the marker names (protein identifiers) and the feature names of the <em>MSnSet</em>. See <em>?addMarkers</em> for examples.</p>
</div>
<div id="data-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-processing" class="anchor" aria-hidden="true"></a>Data processing</h2>
<p>The quantitation data obtained in the supplementary file is normalised to the sum of intensities of each protein; the sum of fraction quantitation for each protein equals 1 (considering rounding errors). This can quickly be verified by computing the row sums of the expression data.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.9990  0.9999  1.0000  1.0000  1.0001  1.0010</code></pre>
<p>The <code>normalise</code> method (also available as <code>normalize</code>) from the <em><a href="https://bioconductor.org/packages/3.13/MSnbase" class="external-link">MSnbase</a></em> package can be used to obtain relative quantitation data, as illustrated below on another iTRAQ test data set, available from <em><a href="https://bioconductor.org/packages/3.13/MSnbase" class="external-link">MSnbase</a></em>. Several normalisation <code>methods</code> are available and described in <code><a href="https://lgatto.github.io/MSnbase/reference/normalise-methods.html" class="external-link">?normalise</a></code>. For many algorithms, including classifiers in general and support vector machines in particular, it is important to properly per-process the data. Centering and scaling of the data is also available with the <code>scale</code> method.</p>
<p>In the code chunk below, we first create a test <em>MSnSet</em> instance<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> and illustrate the effect of <code>normalise(..., method = "sum")</code>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## create a small illustrative test data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span>
<span class="va">itraqdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/quantify-methods.html" class="external-link">quantify</a></span><span class="op">(</span><span class="va">itraqdata</span>, method <span class="op">=</span> <span class="st">"trap"</span>,
              reporters <span class="op">=</span> <span class="va">iTRAQ4</span><span class="op">)</span>
<span class="co">## the quantification data</span>
<span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117
## X1   1347.6158  2247.3097  3927.6931  7661.1463
## X10   739.9861   799.3501   712.5983   940.6793
## X11 27638.3582 33394.0252 32104.2879 26628.7278</code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">itraqdata</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max.      NA's 
##     59.06   5638.09  15344.43  38010.87  42256.61 305739.04         1</code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## normalising to the sum of feature intensitites</span>
<span class="va">itraqnorm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/normalise-methods.html" class="external-link">normalise</a></span><span class="op">(</span><span class="va">itraqdata</span>, method <span class="op">=</span> <span class="st">"sum"</span><span class="op">)</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/pSet-class.html" class="external-link">processingData</a></span><span class="op">(</span><span class="va">itraqnorm</span><span class="op">)</span></code></pre></div>
<pre><code>## - - - Processing information - - -
## Data loaded: Wed May 11 18:54:39 2011 
## Updated from version 0.3.0 to 0.3.1 [Fri Jul  8 20:23:25 2016] 
## iTRAQ4 quantification by trapezoidation: Sat May  1 09:28:33 2021 
## Normalised (sum): Sat May  1 09:28:33 2021 
##  MSnbase version: 1.1.22</code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">itraqnorm</span><span class="op">)</span>, n <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>##     iTRAQ4.114 iTRAQ4.115 iTRAQ4.116 iTRAQ4.117
## X1  0.08875373  0.1480074  0.2586772  0.5045617
## X10 0.23178064  0.2503748  0.2232022  0.2946424
## X11 0.23077081  0.2788287  0.2680598  0.2223407</code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">itraqnorm</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
##       1       1       1       1       1       1       1</code></pre>
<p>Note above how the processing undergone by the <em>MSnSet</em> instances <code>itraqdata</code> and <code>itraqnorm</code> is stored in another such specific sub-container, the <code>processingData</code> slot.</p>
<p>The different features (proteins in the <code>tan2009r1</code> data above, but these could also represent peptides or MS<span class="math inline">\(^2\)</span> spectra) are characterised by unique names. These can be retrieved with the <code>featureNames</code> function.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "P20353" "P53501" "Q7KU78" "P04412" "Q7KJ73" "Q7JZN0"</code></pre>
<p>If we look back at section @ref(sec:csv), we see that these have been automatically assigned using the first columns in the <code>exprsFile.csv</code> and <code>fdataFile.csv</code> files. It is thus crucial for these respective first columns to be identical. Similarly, the sample names can be retrieved with <code>sampleNames(tan2009r1)</code>.</p>
</div>
</div>
<div id="sec:viz" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:viz" class="anchor" aria-hidden="true"></a>Data visualisation</h1>
<p>The following sections will focus on two closely related aspects, data visualisation and data analysis (i.e. organelle assignments). Data visualisation is used in the context on quality control, to convince ourselves that the data displays the expected properties so that the output of further processing can be trusted. Visualising results of the localisation prediction is also essential, to control the validity of these results, before proceeding with orthogonal (and often expensive) <em>dry</em> or <em>wet</em> validation.</p>
<div id="sec:profplot" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:profplot" class="anchor" aria-hidden="true"></a>Profile plots</h2>
<p>The underlying principle of gradient approaches is that we have separated organelles along the gradient and by doing so, generated organelle-specific protein distributions along the gradient fractions. The most natural visualisation is shown on figure @ref(fig:plotdist1), obtained using the sub-setting functionality of <em>MSnSet</em> instances and the <code>plotDist</code> function, as illustrated below.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## indices of the mito markers</span>
<span class="va">j</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">$</span><span class="va">markers.orig</span> <span class="op">==</span> <span class="st">"mitochondrion"</span><span class="op">)</span>
<span class="co">## indices of all proteins assigned to the mito</span>
<span class="va">i</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">$</span><span class="va">PLSDA</span> <span class="op">==</span> <span class="st">"mitochondrion"</span><span class="op">)</span>
<span class="fu"><a href="../reference/plotDist.html">plotDist</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span>,
     markers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">[</span><span class="va">j</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plotdist1-1.png" alt="Distribution of protein intensities along the fractions of the separation gradient for 4 organelles: mitochondrion (red), ER/Golgi (blue, ER markers and green, Golgi markers) and plasma membrane (purple)." width="700"><p class="caption">
Distribution of protein intensities along the fractions of the separation gradient for 4 organelles: mitochondrion (red), ER/Golgi (blue, ER markers and green, Golgi markers) and plasma membrane (purple).
</p>
</div>
</div>
<div id="sec:dendro" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:dendro" class="anchor" aria-hidden="true"></a>Sub-cellular cluster dendrogram</h2>
<p>To gain a quick overview of the distance/similarity between the sub-cellular clusters, it can useful to compare average marker profiles, rather than all profiles, as presented in the profile plots above. The <code>mrkHClust</code> calculates average class profiles and generates the resulting dendrogram.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/mrkHClust.html">mrkHClust</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/dendro-1.png" alt="Hierarchical clustering. Average distance between organelle classes." width="700"><p class="caption">
Hierarchical clustering. Average distance between organelle classes.
</p>
</div>
<p>On figure @ref(fig:dendro), we see that the lysosome and the ribosome 60S are separated by the smallest distance. The advantage of this representation is that it provides a quick snapshot of the average similarity between organelles using the complete profiles (as opposed to a PCA plot, discussed in the next section). The main drawback is that it ignores any variability in individual markers (cluster thighness). It is however a good guide for a more thorough exploration of the data, as described in the next sections.</p>
<p>Note the colours of the labels on the dendrogram (figure @ref(fig:dendro)), which match the colours used to annotate PCA plots, described in the next section (figure @ref(fig:plot2d). These colours are defined at the session level (see <code>getStockcol</code> and <code>setStockcol</code>) and re-used throughout <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> for consistent annotation.</p>
</div>
<div id="sec:avrprofplot" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:avrprofplot" class="anchor" aria-hidden="true"></a>Average organelle class profile plot</h2>
<p>We can also visualise the average organelle class profiles generated by <code>mrkConsProfiles</code> using <code>plotConsProfiles</code>. We can optionally order the organelle classes on the y-axis according to the heirachical clustering from <code>mrkHClust</code>.</p>
<p>See <code><a href="../reference/mrkHClust.html">?mrkHClust</a></code> for more details</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## histogram</span>
<span class="va">hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrkHClust.html">mrkHClust</a></span><span class="op">(</span><span class="va">tan2009r1</span>, plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="co">## order of markers according to histogram</span>
<span class="va">mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getMarkerClasses.html">getMarkerClasses</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span>
<span class="va">m_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">mm</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/order.dendrogram.html" class="external-link">order.dendrogram</a></span><span class="op">(</span><span class="va">hc</span><span class="op">)</span><span class="op">]</span>

<span class="co">## average marker profile</span>
<span class="va">fmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mrkConsProfiles.html">mrkConsProfiles</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotConsProfiles.html">plotConsProfiles</a></span><span class="op">(</span><span class="va">fmat</span>, order<span class="op">=</span><span class="va">m_order</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plotavrprof-1.png" alt="Average organelle class profiles. Protein intensity indicated by colour. Organelle classes ordered by hierarchical clustering" width="700"><p class="caption">
Average organelle class profiles. Protein intensity indicated by colour. Organelle classes ordered by hierarchical clustering
</p>
</div>
</div>
<div id="sec:pcalot" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:pcalot" class="anchor" aria-hidden="true"></a>Dimensionality reduction</h2>
<p>Alternatively, we can combine all organelle groups in one single 2 dimensional figure by applying a dimensionality reduction technique such as Principal Component Analysis (PCA) using the <code>plot2D</code> function (see figure @ref(fig:plot2d)). The protein profile vectors are summarised into 2 values that can be visualised in two dimensions, so that the variability in the data will be maximised along the first principal component (PC1). The second principal component (PC2) is then chosen as to be orthogonal to PC1 while explaining as much variance in the data as possible, and so on for PC3, PC4, etc.</p>
<p>Using a PCA representation to visualise a spatial proteomics experiment, we can easily plot all the proteins on the same figure as well a many sub-cellular clusters. These clusters are defined in a feature meta-data column (slot <code>featureData</code>, accessed as a <code>data.frame</code> with the <code>fData</code> function) that is declared with the <code>fcol</code> argument (default is <code>"markers"</code> which contains the most current known markers for this experiment, while the original markers published with the data can be found in the slot <code>"markers.orig"</code>).</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers"</span><span class="op">)</span>
<span class="fu"><a href="../reference/addLegend.html">addLegend</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers"</span>, cex <span class="op">=</span> <span class="fl">.7</span>,
      where <span class="op">=</span> <span class="st">"bottomright"</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plot2d-1.png" alt="PCA plot. Representation of the proteins of `tan2009r1` after reduction of the 4 reporter quantitation data to 2 principal components." width="700"><p class="caption">
PCA plot. Representation of the proteins of <code>tan2009r1</code> after reduction of the 4 reporter quantitation data to 2 principal components.
</p>
</div>
<p>As the default value for the <code>fcol</code> argument is <code>"markers"</code>, it is not necessary to specify it. It is however mandatory to specify other annotation feature variables, such as to visualise the set of markers described in the original publication.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers.orig"</span><span class="op">)</span>
<span class="fu"><a href="../reference/addLegend.html">addLegend</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers.orig"</span>, where <span class="op">=</span> <span class="st">"bottomright"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plot2dorg-1.png" alt="PCA plot. Reduced set of markers for the `tan2009r1` data projected onto 2 principal components." width="700"><p class="caption">
PCA plot. Reduced set of markers for the <code>tan2009r1</code> data projected onto 2 principal components.
</p>
</div>
<p>It is also possible to visualise the data along 3 dimensions using the <em>plot3D</em> function, which works like the 2 dimension version (<a href="#fig:plot3D">figure below</a>). The resulting figure can be rotated by the user.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/MSmap-class.html" class="external-link">plot3D</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Figures/plot3D.png" id="fig:plot3D" alt=""><p class="caption">Snapshot of the 3-dimensional PCA plot. The <code>tan2009r1</code> data is represented along the first 3 principal components.</p>
</div>
<p>As can be seen on the figures @ref(fig:plot2d), @ref(fig:plot2dorg) and on the <a href="#fig:plot3D">3D plot above</a>, we indicate on the axis labels the percentage of total variance explained by the individual PCs. It is not unusual to reach over 75% along the first two PCs, even for experiments with several tens of fractions. One can calculate this information for all PCs by setting <code>method = "scree"</code> in <code>plot2D</code>. On figure @ref(fig:scree), we see that the four PCs in the <code>tan2009r1</code> data account for 58.53, 29.96, 11.52 and 0 percent of the total variance.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">tan2009r1</span>, method <span class="op">=</span> <span class="st">"scree"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/scree-1.png" alt="Percentage of variance explained." width="700"><p class="caption">
Percentage of variance explained.
</p>
</div>
<p>A variety of dimensionality reduction methods are available in <code>plot2D</code>: PCA, MDS, kpca, lda, t-SNE, nipals, hexbin, none, scree. Except for <code>scree</code> (see above) and <code>none</code> (no data transformation, which can be useful when the data is already transformed and only needs to be plotted), these can used to produce visualisation of the data in two dimensions. Two are worth some discussion here; the readers are redirected to the manual page for more details.</p>
<p>Linear discriminant analysis (LDA) will project the protein occupancy profiles in a new set of dimensions using as a criterion the separation of marker classes by maximising the between class variance to the within class variance ratio. As opposed to the <em>unsupervised</em> PCA, the <em>supervised</em> LDA should not be used as an to explore the data or as a quality control, but can be useful to assess if one or more organelles have been preferentially separated.</p>
<p>The t-Distributed Stochastic Neighbour Embedding (t-SNE)<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> <span class="citation">(Maaten and Hinton 2008)</span> is widely applied in many areas in computational biology and more generally field that need to visualise high-dimensional data. The t-SNE method is non-linear, and will emphasise separation of different features while grouping features with similar profiles. In addition, different transformations are applied on different regions leading to plots that can substantially differ from a PCA plot. As a result, proximity in two dimensions and tightness of the clusters can’t be related to these quantities in the original data. See <em>How to Use t-SNE Effectively</em><a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a> for a useful non-technical introduction.</p>
<p>The results of the algorithm crucially depend on the values of its input parameters, in particular the <em>perplexity</em>, which balances global and local aspects of the data. The suggested range of value ranges from 5 to 50, and should not be greater than the number of data points (which we can assume not to be the case in modern spatial proteomics experiments). Below, we test the effect of this parameter along the suggested range, including the default value of 30, at which the algorithm converges.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">50</span>, <span class="fl">15</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">HEK293T2011</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">HEK293T2011</span>, main <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">perps</span>,
       <span class="kw">function</span><span class="op">(</span><span class="va">perp</span><span class="op">)</span> <span class="op">{</span>
       <span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">HEK293T2011</span>, method <span class="op">=</span> <span class="st">"t-SNE"</span>,
          methargs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>perplexity <span class="op">=</span> <span class="va">perp</span><span class="op">)</span><span class="op">)</span>
       <span class="fu"><a href="https://rdrr.io/r/graphics/title.html" class="external-link">title</a></span><span class="op">(</span>main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"t-SNE, perplexity"</span>, <span class="va">perp</span><span class="op">)</span><span class="op">)</span>
       <span class="op">}</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Figures/tsne.png" id="fig:tsne" alt=""><p class="caption">Effect of t-SNE’s perplexity parameter on the human <code>HEK293T2011</code> data.</p>
</div>
<p>Other parameters that can effect the results are the number of iterations and the learning rate epsilon.</p>
<p>The t-SNE algorithm takes much more time to complete than the other available methods. In such cases, saving the results and re-plotting with method <code>none</code> can be useful (see <code><a href="../reference/plot2D.html">?plot2D</a></code>). In the case of this document, the <a href="#fig:tsne">figure above</a>, was pre-generated rather than computed upon compilation.</p>
</div>
<div id="features-of-interest" class="section level2">
<h2 class="hasAnchor">
<a href="#features-of-interest" class="anchor" aria-hidden="true"></a>Features of interest</h2>
<p>In addition to highlighting sub-cellular niches as coloured clusters on the PCA plot, it is also possible to define some arbitrary <em>features of interest</em> that represent, for example, proteins of a particular pathway or a set of interaction partners. Such sets of proteins are recorded as <em>FeaturesOfInterest</em> instances, as illstrated below (using the ten first features of our experiment):</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">foi1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/FeaturesOfInterest-class.html" class="external-link">FeaturesOfInterest</a></span><span class="op">(</span>description <span class="op">=</span> <span class="st">"Feats of interest 1"</span>,
               fnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/description.html" class="external-link">description</a></span><span class="op">(</span><span class="va">foi1</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Feats of interest 1"</code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/FeaturesOfInterest-class.html" class="external-link">foi</a></span><span class="op">(</span><span class="va">foi1</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "P20353" "P53501" "Q7KU78" "P04412" "Q7KJ73" "Q7JZN0" "Q7KLV9" "Q9VM65"
##  [9] "Q9VCK0" "Q9VIU7"</code></pre>
<p>Several such features of interest can be combined into collections:</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">foi2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/FeaturesOfInterest-class.html" class="external-link">FeaturesOfInterest</a></span><span class="op">(</span>description <span class="op">=</span> <span class="st">"Feats of interest 2"</span>,
               fnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureNames.html" class="external-link">featureNames</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">[</span><span class="fl">880</span><span class="op">:</span><span class="fl">888</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">foic</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/FeaturesOfInterest-class.html" class="external-link">FoICollection</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">foi1</span>, <span class="va">foi2</span><span class="op">)</span><span class="op">)</span>
<span class="va">foic</span></code></pre></div>
<pre><code>## A collection of 2 features of interest.</code></pre>
<p><em>FeaturesOfInterest</em> instances can now be overlaid on the PCA plot with the <code>highlightOnPlot</code> function. The <code>highlightOnPlot3D</code> can be used to overlay data onto a 3 dimensional figure produced by <code>plot3D</code>.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"PLSDA"</span><span class="op">)</span>
<span class="fu"><a href="../reference/addLegend.html">addLegend</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"PLSDA"</span>,
      where <span class="op">=</span> <span class="st">"bottomright"</span>,
      cex <span class="op">=</span> <span class="fl">.7</span><span class="op">)</span>
<span class="fu"><a href="../reference/highlightOnPlot.html">highlightOnPlot</a></span><span class="op">(</span><span class="va">tan2009r1</span>, <span class="va">foi1</span>,
        col <span class="op">=</span> <span class="st">"black"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="../reference/highlightOnPlot.html">highlightOnPlot</a></span><span class="op">(</span><span class="va">tan2009r1</span>, <span class="va">foi2</span>,
        col <span class="op">=</span> <span class="st">"purple"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topright"</span>, <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FoI 1"</span>, <span class="st">"FoI 2"</span><span class="op">)</span>,
       bty <span class="op">=</span> <span class="st">"n"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"purple"</span><span class="op">)</span>,
       pch <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plotfoi-1.png" alt="Adding features of interest on a PCA plot." width="700"><p class="caption">
Adding features of interest on a PCA plot.
</p>
</div>
<p>See <code><a href="https://lgatto.github.io/MSnbase/reference/FeaturesOfInterest-class.html" class="external-link">?FeaturesOfInterest</a></code> and <code><a href="../reference/highlightOnPlot.html">?highlightOnPlot</a></code> for more details.</p>
</div>
<div id="sec:gui" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:gui" class="anchor" aria-hidden="true"></a>Interactive visualisation</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.13/pRolocGUI" class="external-link">pRolocGUI</a></em> application allows one to explore the spatial proteomics data using an interactive, web-based  interface <span class="citation">(RStudio and Inc. 2014)</span>. The package is available from Bioconductor and can be installed and started as follows:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"pRolocGUI"</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"pRolocGUI"</span><span class="op">)</span>
<span class="fu">pRolocVis</span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="Figures/Screenshot_PCA2.png" id="fig:gui" alt=""><p class="caption">Screenshot of the <em><a href="https://bioconductor.org/packages/3.13/pRolocGUI" class="external-link">pRolocGUI</a></em> interface. The GUI is also available as an online app for the hyperLOPIT experiment from <span class="citation">(Christoforou et al. 2016)</span> at <a href="https://lgatto.shinyapps.io/christoforou2015/" class="external-link uri">https://lgatto.shinyapps.io/christoforou2015/</a>.</p>
</div>
<p>More details are available in the vignette that can be started from the application by clicking on any of the question marks, by starting the vignette from R with <code>vignette("pRolocGUI")</code> or can be accessed online (<a href="http://bioconductor.org/packages/devel/bioc/vignettes/pRolocGUI/inst/doc/pRolocGUI.html" class="external-link uri">http://bioconductor.org/packages/devel/bioc/vignettes/pRolocGUI/inst/doc/pRolocGUI.html</a>).</p>
</div>
</div>
<div id="sec:qsep" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:qsep" class="anchor" aria-hidden="true"></a>Assessing sub-cellular resolution</h1>
<p>The sub-cellular resolution of a spatial proteomics experiment, i.e. the quantitation of how well the respective sub-cellular niches are separated, can be computed with the <code>QSep</code> function. Briefly, this function compares, for each pairs of sub-cellular niches, the ratio between the the average Euclidean distance between niche <em>i</em> and <em>j</em> and the the average within distance of cluster <em>j</em>. A large ratio indicates that <em>i</em> and <em>j</em> are well separated with respect to the thighness of cluster <em>j</em>. The larger the distances, the better the spatial proteomics experiment.</p>
<p>Below, we calulate and visualise the <em>QSep</em> distances for the <code>hyperLOPIT2015</code> data:</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/lgatto/pRolocdata" class="external-link">"pRolocdata"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">hyperLOPIT2015</span><span class="op">)</span>

<span class="co">## Create the object and get a summary</span>
<span class="va">hlq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/QSep-class.html">QSep</a></span><span class="op">(</span><span class="va">hyperLOPIT2015</span><span class="op">)</span>
<span class="va">hlq</span></code></pre></div>
<pre><code>## Object of class 'QSep'.
##  Data: hyperLOPIT2015 
##  With 14 sub-cellular clusters.</code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GenRegRes-class.html">levelPlot</a></span><span class="op">(</span><span class="va">hlq</span><span class="op">)</span></code></pre></div>
<p><img src="v01-pRoloc-tutorial_files/figure-html/qsep-1.png" width="700"></p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/plot-methods.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">hlq</span><span class="op">)</span></code></pre></div>
<p><img src="v01-pRoloc-tutorial_files/figure-html/qsep-2.png" width="700"></p>
<p>See <a href="https://doi.org/10.1101/377630" class="external-link">Assessing sub-cellular resolution in spatial proteomics experiments</a> <span class="citation">(Gatto, Breckels, and Lilley 2018)</span> for details, including a large meta-analysis on 29 different spatial proteomics experiments.</p>
</div>
<div id="sec:analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:analysis" class="anchor" aria-hidden="true"></a>Data analysis</h1>
<p>Classification of proteins, i.e. assigning sub-cellular localisation to proteins, is the main aspect of the present data analysis. The principle is the following and is, in its basic form, a 2 step process. First, an algorithm learns from the known markers that are shown to him and models the data space accordingly. This phase is also called the training phase. In the second phase, un-labelled proteins, i.e. those that have not been labelled as resident of any organelle, are matched to the model and assigned to a group (an organelle). This 2 step process is called machine learning (ML), because the computer (machine) learns by itself how to recognise instances that possess certain characteristics and classifies them without human intervention. That does however not mean that results can be trusted blindly.</p>
<p>In the above paragraph, we have defined what is called supervised ML, because the algorithm is presented with some know instances from which it learns (see section @ref(sec:sml)). Alternatively, un-supervised ML does not make any assumptions about the group memberships, and uses the structure of the data itself to defined sub-groups (see section @ref(sec:usml)). It is of course possible to classify data based on labelled and unlabelled data. This extension of the supervised classification problem described above is called semi-supervised learning. In this case, the training data consists of both labelled and unlabelled instances with the obvious goal of generating a better classifier than would be possible with the labelled data only. The <code>phenoDisco</code> algorithm, will be illustrated in that context (section @ref(sec:ssml)).</p>
<div id="sec:usml" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:usml" class="anchor" aria-hidden="true"></a>Unsupervised ML</h2>
<p>The <code>plot2D</code> can also be used to visualise the data on a PCA plot omitting any marker definitions, as shown on figure @ref(fig:plot2dnull). This approach avoids any bias towards marker definitions and concentrate on the data and its underlying structure itself.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plot2dnull-1.png" alt="Plain PCA representation of the `tan2009r1` data." width="700"><p class="caption">
Plain PCA representation of the <code>tan2009r1</code> data.
</p>
</div>
<p>Alternatively, <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> also gives access to <em><a href="https://bioconductor.org/packages/3.13/MLInterfaces" class="external-link">MLInterfaces</a></em>’s <code>MLean</code> unified interface for, among others, unsupervised approaches using k-means (figure @ref(fig:plotKmeans)), hierarchical (figure @ref(fig:plotHclust)) or partitioning around medoids (figure @ref(fig:plotPam)), clustering.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">kcl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MLInterfaces/man/MLearn-new.html" class="external-link">MLearn</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">tan2009r1</span>,  <span class="va">kmeansI</span>, centers<span class="op">=</span><span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/plot-methods.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">kcl</span>, <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plotKmeans-1.png" alt="k-means clustering on the `tan2009r1` data." width="700"><p class="caption">
k-means clustering on the <code>tan2009r1</code> data.
</p>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hcl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MLInterfaces/man/MLearn-new.html" class="external-link">MLearn</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">tan2009r1</span>,
          <span class="fu"><a href="https://rdrr.io/pkg/MLInterfaces/man/MLearn-new.html" class="external-link">hclustI</a></span><span class="op">(</span>distFun <span class="op">=</span> <span class="va">dist</span>,
              cutParm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/plot-methods.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">hcl</span>, labels <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plotHclust-1.png" alt="Hierarchical clustering on the `tan2009r1` data." width="700"><p class="caption">
Hierarchical clustering on the <code>tan2009r1</code> data.
</p>
</div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pcl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MLInterfaces/man/MLearn-new.html" class="external-link">MLearn</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">.</span>, <span class="va">tan2009r1</span>,  <span class="fu"><a href="https://rdrr.io/pkg/MLInterfaces/man/MLearn-new.html" class="external-link">pamI</a></span><span class="op">(</span><span class="va">dist</span><span class="op">)</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/plot-methods.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pcl</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/exprs.html" class="external-link">exprs</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/plotPam-1.png" alt="Partitioning around medoids on the `tan2009r1` data." width="700"><p class="caption">
Partitioning around medoids on the <code>tan2009r1</code> data.
</p>
</div>
</div>
<div id="sec:sml" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:sml" class="anchor" aria-hidden="true"></a>Supervised ML</h2>
<p>In this section, we show how to use <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> to run a typical supervised ML analysis. Several ML methods are available, including k-nearest neighbour (knn), partial least square discriminant analysis (plsda), random forest (rf), support vector machines (svm), The detailed description of each method is outside of the scope of this document. We will use support vector machines to illustrate a typical pipeline and the important points that should be paid attention to. These points are equally valid and work, from a <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> user perspective, exactly the same for the other approaches.</p>
<div id="classification-algorithm-parameters-optimisation" class="section level3">
<h3 class="hasAnchor">
<a href="#classification-algorithm-parameters-optimisation" class="anchor" aria-hidden="true"></a>Classification algorithm parameters optimisation</h3>
<p>Before actually generating a model on the new markers and classifying unknown residents, one has to take care of properly setting the model parameters. Wrongly set parameters can have a very negative impact on classification performance. To do so, we create testing (to model) and training (to predict) subsets using known residents, i.e. marker proteins. By comparing observed and expected classification prediction, we can assess how well a given model works using the macro F1 score (see below). This procedure is repeated for a range of possible model parameter values (this is called a grid search), and the best performing set of parameters is then used to construct a model on all markers and predict un-labelled proteins. For this parameter optimisation procedure to perform well and produce useful results, it is essential to run it with a reasonable amount of markers. In our experience, 15 such marker proteins are necessary.</p>
<p>Model accuracy is evaluated using the F1 score, <span class="math inline">\(F1 = 2 ~ \frac{precision \times recall}{precision + recall}\)</span>, calculated as the harmonic mean of the precision (<span class="math inline">\(precision = \frac{tp}{tp+fp}\)</span>, a measure of <em>exactness</em> – returned output is a relevant result) and recall (<span class="math inline">\(recall=\frac{tp}{tp+fn}\)</span>, a measure of <em>completeness</em> – indicating how much was missed from the output). What we are aiming for are high generalisation accuracy, i.e high <span class="math inline">\(F1\)</span>, indicating that the marker proteins in the test data set are consistently correctly assigned by the algorithms.</p>
<p>In order to evaluate how well a classifier performs on profiles it was not exposed to during its creation, we implemented the following schema. Each set of marker protein profiles, i.e. labelled with known organelle association, are separated into <em>training</em> and <em>test/validation</em> partitions by sampling 80% of the profile corresponding to each organelle (i.e. stratified) without replacement to form the training partition <span class="math inline">\(S_{tr}\)</span> with the remainder becoming the test/validation partition <span class="math inline">\(S_{ts}\)</span>. The svm regularisation parameter <span class="math inline">\(C\)</span> and Gaussian kernel width <span class="math inline">\(sigma\)</span> are selected using a further round of stratified five-fold cross-validation on each training partition. All pairs of parameters <span class="math inline">\((C_i, sigma_j)\)</span> under consideration are assessed using the macro F1 score and the pair that produces the best performance is subsequently employed in training a classifier on all training profiles <span class="math inline">\(S_{tr}\)</span> prior to assessment on all test/validation profiles <span class="math inline">\(S_{ts}\)</span>. This procedure is repeated <span class="math inline">\(N\)</span> times (in the example below 10) in order to produce <span class="math inline">\(N\)</span> macro F1 estimated generalisation performance values (figures @ref(fig:params1) and @ref(fig:params2)). This procedure is implemented in the <code>svmOptimisation</code>. See <code><a href="../reference/svmOptimisation.html">?svmOptimisation</a></code> for details, in particular the range of <span class="math inline">\(C\)</span> and <span class="math inline">\(sigma\)</span> parameters and how the relevant feature variable is defined by the <code>fcol</code> parameters, which defaults to <code>"markers"</code>.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svmOptimisation.html">svmOptimisation</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers.orig"</span>,
              times <span class="op">=</span> <span class="fl">100</span>, xval <span class="op">=</span> <span class="fl">5</span>,
              verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>In the interest of time, the optimisation is not executed but loaded with</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRoloc"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"params.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params</span></code></pre></div>
<pre><code>## Object of class "GenRegRes"
## Algorithm: svm 
## Hyper-parameters:
##  cost: 0.0625 0.125 0.25 0.5 1 2 4 8 16
##  sigma: 0.01 0.1 1 10 100 1000
## Design:
##  Replication: 10 x 5-fold X-validation
##  Partitioning: 0.2/0.8 (test/train)
## Results
##  macro F1:
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.8889  0.8889  1.0000  0.9556  1.0000  1.0000 
##  best sigma: 0.1 1   
##  best cost: 0.5 1</code></pre>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/plot-methods.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">params</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/params1-1.png" alt="Assessing parameter optimisation. Here, we see the respective distributions of the 10 macro F1 scores for the best cost/sigma parameter pairs. See also the output of *f1Count* in relation to this plot." width="700"><p class="caption">
Assessing parameter optimisation. Here, we see the respective distributions of the 10 macro F1 scores for the best cost/sigma parameter pairs. See also the output of <em>f1Count</em> in relation to this plot.
</p>
</div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GenRegRes-class.html">levelPlot</a></span><span class="op">(</span><span class="va">params</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/params2-1.png" alt="Assessing parameter optimisation. Visualisation of the averaged macro F1 scores, for the full range of parameter values." width="700"><p class="caption">
Assessing parameter optimisation. Visualisation of the averaged macro F1 scores, for the full range of parameter values.
</p>
</div>
<p>In addition to the plots on figures @ref(fig:params1) and @ref(fig:params2), <code>f1Count(params)</code> returns, for each combination of parameters, the number of best (highest) F1 observations. One can use <code>getParams</code> to see the default set of parameters that are chosen based on the executed optimisation. Currently, the first best set is automatically extracted, and users are advised to critically assess whether this is the most wise choice.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GenRegRes-class.html">f1Count</a></span><span class="op">(</span><span class="va">params</span><span class="op">)</span></code></pre></div>
<pre><code>##     0.5 1
## 0.1   1 0
## 1    NA 5</code></pre>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/GenRegRes-class.html">getParams</a></span><span class="op">(</span><span class="va">params</span><span class="op">)</span></code></pre></div>
<pre><code>## sigma  cost 
##   0.1   0.5</code></pre>
<p><strong>Important</strong> It is essential to emphasise that the accuracy scores obtained during parameter optimisation are only a reflection of the classification performance on a set of distinct and ideally separated spatial clusters. Here, we assume that the data is characterised by good separation of the various spatial niches which are reflected by sub-cellular markers. Quality control of the data and the markers using the visualisation described in section @ref(sec:viz) is essential and subsequent analyses are doomed to fail in the absence of such separation.</p>
<p>These classification scores are not representative of the reliability of the final classification (described in section @ref(sec:sml)), in particular along the boundaries separating the different sub-cellular niches. High scores at the optimisation stage are a requirement to proceed with the analysis, but are by no means indicative of the false positive rate in the final sub-cellular assignment of non-marker proteins.</p>
</div>
<div id="classification" class="section level3">
<h3 class="hasAnchor">
<a href="#classification" class="anchor" aria-hidden="true"></a>Classification</h3>
<p>We can now re-use the result from our parameter optimisation (a best cost/sigma pair is going to be automatically extracted, using the <code>getParams</code> method, although it is possible to set them manually), and use them to build a model with all the marker proteins and predict unknown residents using the <em>svmClassification</em> function (see the manual page for more details). By default, the organelle markers will be defined by the <code>"markers"</code> feature variables (and can be defined by the <code>fcol</code> parameter) e.g. here we use the original markers in <code>"markers.orig"</code> as a use case. New feature variables containing the organelle assignments and assignment probabilities, called scores hereafter, are automatically added to the <code>featureData</code> slot; in this case, using the <code>svm</code> and <code>svm.scores</code> labels.</p>
<p><strong>Important</strong> The calculation of the classification probabilities is dependent on the classification algorithm. These probabilities are not to be compared across algorithms; they do <strong>not</strong> reflect any <strong>biologically relevant</strong> sub-cellular localisation probability but rather an algorithm-specific classification confidence score.}</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## manual setting of parameters</span>
<span class="va">svmres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svmClassification.html">svmClassification</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers.orig"</span>,
                sigma <span class="op">=</span> <span class="fl">1</span>, cost <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## using default best parameters</span>
<span class="va">svmres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svmClassification.html">svmClassification</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"markers.orig"</span>,
                assessRes <span class="op">=</span> <span class="va">params</span><span class="op">)</span>
<span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/pSet-class.html" class="external-link">processingData</a></span><span class="op">(</span><span class="va">svmres</span><span class="op">)</span></code></pre></div>
<pre><code>## - - - Processing information - - -
## Added markers from  'mrk' marker vector. Thu Jul 16 22:53:44 2015 
## Performed svm prediction (sigma=0.1 cost=0.5) Sat May  1 09:28:41 2021 
##  MSnbase version: 1.17.12</code></pre>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fvarLabels</a></span><span class="op">(</span><span class="va">svmres</span><span class="op">)</span>, <span class="fl">4</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "markers"    "markers.tl" "svm"        "svm.scores"</code></pre>
<p>The original markers, classification results and scores can be accessed with the <code>fData</code> accessor method, e.g. <code>fData(svmres)$svm</code> or <code>fData(svmres)$svm.scores</code>. Two helper functions, <code>getMarkers</code> and <code>getPredictions</code> are available and add some level of automation and functionality, assuming that the default feature labels are used. Both (invisibly) return the corresponding feature variable (the markers or assigned classification) and print a summary table. The <code>fcol</code> parameter must be specified for <code>getPredictions</code>. It is also possible to defined a classification probability below which classifications are set to <code>"unknown"</code>.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPredictions.html">getPredictions</a></span><span class="op">(</span><span class="va">svmres</span>, fcol <span class="op">=</span> <span class="st">"svm"</span><span class="op">)</span></code></pre></div>
<pre><code>## ans
##  Cytoskeleton            ER         Golgi      Lysosome mitochondrion 
##             7           241            39             8           219 
##       Nucleus    Peroxisome            PM    Proteasome  Ribosome 40S 
##            21             4           282            15            20 
##  Ribosome 60S 
##            32</code></pre>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">p1</span><span class="op">)</span><span class="op">$</span><span class="va">svm.pred</span>
<span class="va">minprob</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">svmres</span><span class="op">)</span><span class="op">$</span><span class="va">svm.scores</span><span class="op">)</span>
<span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getPredictions.html">getPredictions</a></span><span class="op">(</span><span class="va">svmres</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>, t <span class="op">=</span> <span class="va">minprob</span><span class="op">)</span></code></pre></div>
<pre><code>## ans
##  Cytoskeleton            ER         Golgi      Lysosome mitochondrion 
##             7           174            21             8           102 
##       Nucleus    Peroxisome            PM    Proteasome  Ribosome 40S 
##            21             4           148            15            20 
##  Ribosome 60S       unknown 
##            32           336</code></pre>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">p2</span><span class="op">)</span><span class="op">$</span><span class="va">svm.pred</span>
<span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></code></pre></div>
<pre><code>##                p2
## p1              Cytoskeleton  ER Golgi Lysosome mitochondrion Nucleus
##   Cytoskeleton             7   0     0        0             0       0
##   ER                       0 174     0        0             0       0
##   Golgi                    0   0    21        0             0       0
##   Lysosome                 0   0     0        8             0       0
##   mitochondrion            0   0     0        0           102       0
##   Nucleus                  0   0     0        0             0      21
##   Peroxisome               0   0     0        0             0       0
##   PM                       0   0     0        0             0       0
##   Proteasome               0   0     0        0             0       0
##   Ribosome 40S             0   0     0        0             0       0
##   Ribosome 60S             0   0     0        0             0       0
##                p2
## p1              Peroxisome  PM Proteasome Ribosome 40S Ribosome 60S unknown
##   Cytoskeleton           0   0          0            0            0       0
##   ER                     0   0          0            0            0      67
##   Golgi                  0   0          0            0            0      18
##   Lysosome               0   0          0            0            0       0
##   mitochondrion          0   0          0            0            0     117
##   Nucleus                0   0          0            0            0       0
##   Peroxisome             4   0          0            0            0       0
##   PM                     0 148          0            0            0     134
##   Proteasome             0   0         15            0            0       0
##   Ribosome 40S           0   0          0           20            0       0
##   Ribosome 60S           0   0          0            0           32       0</code></pre>
<p>To graphically illustrate the organelle-specific score distributions, we can use a boxplot and plot the scores for the respective predicted svm classes, as shown on figure @ref(fig:predscores). As can be seen, different organelles are characterised by different score distributions. Using a unique threshold (<code>minprob</code> with value 0.78 above) results in accepting 72% of the initial ER predictions and only 47% of the mitochondrion predictions. The <em>getPredictions</em> function also accepts organelle-specific score thresholds. Below, we calculate organelle-specific median scores.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="va">svm.scores</span> <span class="op">~</span> <span class="va">svm</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">svmres</span><span class="op">)</span>,
    ylab <span class="op">=</span> <span class="st">"SVM scores"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="va">minprob</span>, lwd <span class="op">=</span> <span class="fl">2</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/predscores-1.png" alt="Organelle-specific SVM score distributions." width="700"><p class="caption">
Organelle-specific SVM score distributions.
</p>
</div>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/orgQuants.html">orgQuants</a></span><span class="op">(</span><span class="va">svmres</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>, t <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></code></pre></div>
<pre><code>##            ER         Golgi mitochondrion            PM 
##     0.8374491     0.6318175     0.7483629     0.7725315</code></pre>
<p>Using these scores equates to choosing the 50% predictions with highest scores for organelle.</p>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getPredictions.html">getPredictions</a></span><span class="op">(</span><span class="va">svmres</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>, t <span class="op">=</span> <span class="va">ts</span><span class="op">)</span></code></pre></div>
<pre><code>## ans
##  Cytoskeleton            ER         Golgi      Lysosome mitochondrion 
##             7           135            26             8           124 
##       Nucleus    Peroxisome            PM    Proteasome  Ribosome 40S 
##            21             4           158            15            20 
##  Ribosome 60S       unknown 
##            32           338</code></pre>
<pre><code>## MSnSet (storageMode: lockedEnvironment)
## assayData: 888 features, 4 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: X114 X115 X116 X117
##   varLabels: Fractions
##   varMetadata: labelDescription
## featureData
##   featureNames: P20353 P53501 ... P07909 (888 total)
##   fvarLabels: FBgn Protein.ID ... svm.pred (19 total)
##   fvarMetadata: labelDescription
## experimentData: use 'experimentData(object)'
##   pubMedIds: 19317464 
## Annotation:  
## - - - Processing information - - -
## Added markers from  'mrk' marker vector. Thu Jul 16 22:53:44 2015 
## Performed svm prediction (sigma=0.1 cost=0.5) Sat May  1 09:28:41 2021 
## Added svm predictions according to thresholds: ER = 0.84, Golgi = 0.63, mitochondrion = 0.75, PM = 0.77 Sat May  1 09:28:42 2021 
##  MSnbase version: 1.17.12</code></pre>
<p>We can now visualise these results using the plotting functions presented in section @ref(sec:usml), as shown on figure @ref(fig:svmres). We clearly see that besides the organelle marker clusters that have been assigned high confidence members, many other proteins have substantially lower prediction scores.</p>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ptsze</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">svmres</span><span class="op">)</span><span class="op">$</span><span class="va">svm.scores</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>
<span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">svmres</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>, fpch <span class="op">=</span> <span class="st">"markers.orig"</span>,
       cex <span class="op">=</span> <span class="va">ptsze</span><span class="op">)</span>
<span class="fu"><a href="../reference/addLegend.html">addLegend</a></span><span class="op">(</span><span class="va">svmres</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>,
      where <span class="op">=</span> <span class="st">"bottomright"</span>,
      cex <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/svmres-1.png" alt="Representation of the svm prediction on the `tan2009r1` data set. The svm scores have been used to set the point size (`cex` argument; the scores have been transformed to emphasise the extremes). Different symbols (`fpch`) are used to differentiate markers and new assignments." width="700"><p class="caption">
Representation of the svm prediction on the <code>tan2009r1</code> data set. The svm scores have been used to set the point size (<code>cex</code> argument; the scores have been transformed to emphasise the extremes). Different symbols (<code>fpch</code>) are used to differentiate markers and new assignments.
</p>
</div>
</div>
</div>
<div id="sec:bayes" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:bayes" class="anchor" aria-hidden="true"></a>Bayesian generative models</h2>
<p>We also offer generative models that, as opposed to the descriptive classifier presented above, explicitly model the spatial proteomics data. In <code>pRoloc</code>, we probose two models using T-augmented Gaussian mixtures using repectively a Expectration-Maximisation approach to <em>maximum a posteriori</em> estimation of the model parameters (TAGM-MAP), and an MCMC approach (TAGM-MCMC) that enables a proteome-wide uncertainty quantitation. These methods are described in the <em>pRoloc-bayesian</em> vignette. For a details description of the methods and their validation, please refer to <span class="citation">(Crook et al. 2018)</span>.</p>
</div>
<div id="sec:ssml" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:ssml" class="anchor" aria-hidden="true"></a>Semi-supervised ML</h2>
<p>It is obvious that the original set of markers initially used (ER, Golgi, mitochondrion, PM) is not a biologically realistic representation or the organelle diversity. Manually finding markers is however time consuming, as it requires careful verification of the annotation, and possibly critical for the subsequent analysis, as markers are directly used in the training phase of the supervised ML approach.</p>
<p>As can be seen in the PCA plots above, there is inherent structure in the data that can be made use of to automate the detection of new clusters. The <code>phenoDisco</code> algorithm <span class="citation">(Breckels et al. 2013)</span> is an iterative method, that combines classification of proteins to known groups and detection of new clusters. It is available in <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em> though the <code>phenoDisco</code> function.</p>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pdres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phenoDisco.html">phenoDisco</a></span><span class="op">(</span><span class="va">tan2009r1</span>, GS <span class="op">=</span> <span class="fl">10</span>,
                    times <span class="op">=</span> <span class="fl">100</span>, fcol <span class="op">=</span> <span class="st">"PLSDA"</span><span class="op">)</span></code></pre></div>
<p>Again, in the interest of time, <em>phenoDisco</em> is not executed when the vignette is dynamically built. The data object can be located in the <code>extdata</code> direcoty and loaded with:</p>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRoloc"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"pdres.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span></code></pre></div>
<p>The results are also appended to the <code>featureData</code> slot.</p>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://lgatto.github.io/MSnbase/reference/pSet-class.html" class="external-link">processingData</a></span><span class="op">(</span><span class="va">pdres</span><span class="op">)</span></code></pre></div>
<pre><code>## - - - Processing information - - -
## Combined [888,4] and [1,4] MSnSets Wed Feb 13 17:28:54 2013 
## Run phenoDisco using 'PLSDA': Wed Feb 13 17:28:54 2013 
##   with parameters times=100, GS=10, p=0.05, r=1. 
##  MSnbase version: 1.5.13</code></pre>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fvarLabels</a></span><span class="op">(</span><span class="va">pdres</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "PLSDA"   "markers" "pd"</code></pre>
<p>The <code>plot2D</code> function, can, as previously, be utilised to visualise the results, as shown on figure @ref(fig:pdres).</p>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">pdres</span>, fcol <span class="op">=</span> <span class="st">"pd"</span><span class="op">)</span>
<span class="fu"><a href="../reference/addLegend.html">addLegend</a></span><span class="op">(</span><span class="va">pdres</span>, fcol <span class="op">=</span> <span class="st">"pd"</span>, ncol <span class="op">=</span> <span class="fl">2</span>,
      where <span class="op">=</span> <span class="st">"bottomright"</span>,
      cex <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/pdres-1.png" alt="Representation of the phenoDisco prediction and cluster discovery results." width="700"><p class="caption">
Representation of the phenoDisco prediction and cluster discovery results.
</p>
</div>
</div>
<div id="sec:followup" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:followup" class="anchor" aria-hidden="true"></a>Following up on novelty discovery</h2>
<p>The newly discovered phenotypes need to be carefully validated prior to further analysis. Indeed, as the structure of the data is made use of in the discovery algorithm, some might represent peculiar structure in the data and not match with biologically relevant groups. The <code>tan2009r1</code> data has been submitted to a careful <code>phenodisco</code> analysis and validation in <span class="citation">(Breckels et al. 2013)</span>. The results of this new, augmented marker set is available in the <code>pd.markers</code> feature data. These markers represent a combined set of the original markers and validated proteins from the new phenotypes.</p>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getMarkers.html">getMarkers</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"pd.markers"</span><span class="op">)</span></code></pre></div>
<pre><code>## organelleMarkers
##  Cytoskeleton            ER         Golgi      Lysosome mitochondrion 
##             7            20             6             8            14 
##       Nucleus    Peroxisome            PM    Proteasome  Ribosome 40S 
##            20             4            15            11            14 
##  Ribosome 60S       unknown 
##            25           744</code></pre>
<p>The augmented set of markers is now employed to repeat the classification using the support vector machine classifier. We apply a slightly different analysis than described in section @ref(sec:sml). In the code chunks below, we use class specific weights when creating the svm model; the weights are set to be inversely proportional to class frequencies.</p>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">w</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classWeights.html">classWeights</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"pd.markers"</span><span class="op">)</span>
<span class="va">w</span></code></pre></div>
<pre><code>## 
##  Cytoskeleton            ER         Golgi      Lysosome mitochondrion 
##    0.14285714    0.05000000    0.16666667    0.12500000    0.07142857 
##       Nucleus    Peroxisome            PM    Proteasome  Ribosome 40S 
##    0.05000000    0.25000000    0.06666667    0.09090909    0.07142857 
##  Ribosome 60S 
##    0.04000000</code></pre>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">params2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svmOptimisation.html">svmOptimisation</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"pd.markers"</span>,
               times <span class="op">=</span> <span class="fl">10</span>, xval <span class="op">=</span> <span class="fl">5</span>,
               class.weights <span class="op">=</span> <span class="va">w</span>,
               verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>As above, the results are pre-computed and available in the <code>extdata</code> package directory.</p>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"pRoloc"</span><span class="op">)</span>,
      full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"params2.rda"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">fn</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tan2009r1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/svmClassification.html">svmClassification</a></span><span class="op">(</span><span class="va">tan2009r1</span>, <span class="va">params2</span>,
                               class.weights <span class="op">=</span> <span class="va">w</span>,
                               fcol <span class="op">=</span> <span class="st">"pd.markers"</span><span class="op">)</span></code></pre></div>
<p>The data is visualised as described previously, were we use the svm classification a-posteriori probability to set the point size.</p>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ptsze</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData</a></span><span class="op">(</span><span class="va">tan2009r1</span><span class="op">)</span><span class="op">$</span><span class="va">svm.scores</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span>
<span class="fu"><a href="../reference/plot2D.html">plot2D</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>, cex <span class="op">=</span> <span class="va">ptsze</span><span class="op">)</span>
<span class="fu"><a href="../reference/addLegend.html">addLegend</a></span><span class="op">(</span><span class="va">tan2009r1</span>, fcol <span class="op">=</span> <span class="st">"svm"</span>,
      where <span class="op">=</span> <span class="st">"bottomright"</span>,
      ncol <span class="op">=</span> <span class="fl">2</span>, cex <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="v01-pRoloc-tutorial_files/figure-html/pdres2-1.png" alt="Classification results. The results of the second round of classification, using the augmented set of markers obtained using *phenoDisco* as detailed in [@Breckels2013] and a weighted svm classifier." width="700"><p class="caption">
Classification results. The results of the second round of classification, using the augmented set of markers obtained using <em>phenoDisco</em> as detailed in <span class="citation">(Breckels et al. 2013)</span> and a weighted svm classifier.
</p>
</div>
</div>
</div>
<div id="sec:ccl" class="section level1">
<h1 class="hasAnchor">
<a href="#sec:ccl" class="anchor" aria-hidden="true"></a>Conclusions</h1>
<p>This tutorial focuses on practical aspects of organelles proteomics data analysis using <em><a href="https://bioconductor.org/packages/3.13/pRoloc" class="external-link">pRoloc</a></em>. Two important aspects have been illustrates: (1) data generation, manipulation and visualisation and (2) application of contemporary and novel machine learning techniques. Other crucial parts of a full analysis pipeline that were not covered here are raw mass-spectrometry quality control, quantitation, post-analysis and data validation.</p>
<p>Data analysis is not a trivial task, and in general, one can not assume that any off-the-shelf algorithm will perform well. As such, one of the emphasis of the software presented in this document is allowing users to track data processing and critically evaluate the results.</p>
</div>
<div id="acknowledgement" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#acknowledgement" class="anchor" aria-hidden="true"></a>Acknowledgement</h1>
<p>We would like to thank Dr Daniel J.H. Nightingale, Dr Arnoud J. Groen, Dr Claire M. Mulvey and Dr Andy Christoforou for their organelle marker contributions.</p>
</div>
<div id="session-information" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor" aria-hidden="true"></a>Session information</h1>
<p>All software and respective versions used to produce this document are listed below.</p>
<pre><code>## R Under development (unstable) (2021-04-15 r80175)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] class_7.3-18         pRolocdata_1.29.0    pRoloc_1.31.2       
##  [4] BiocParallel_1.25.5  MLInterfaces_1.71.2  cluster_2.1.1       
##  [7] annotate_1.69.2      XML_3.99-0.6         AnnotationDbi_1.53.1
## [10] IRanges_2.25.10      MSnbase_2.17.7       ProtGenerics_1.23.9 
## [13] S4Vectors_0.29.17    mzR_2.25.4           Rcpp_1.0.6          
## [16] Biobase_2.51.0       BiocGenerics_0.37.4  knitr_1.33          
## [19] BiocStyle_2.19.2    
## 
## loaded via a namespace (and not attached):
##   [1] BiocFileCache_1.99.7  systemfonts_1.0.1     plyr_1.8.6           
##   [4] splines_4.1.0         ggplot2_3.3.3         digest_0.6.27        
##   [7] foreach_1.5.1         htmltools_0.5.1.1     viridis_0.6.0        
##  [10] gdata_2.18.0          fansi_0.4.2           magrittr_2.0.1       
##  [13] memoise_2.0.0         doParallel_1.0.16     mixtools_1.2.0       
##  [16] limma_3.47.12         recipes_0.1.16        Biostrings_2.59.2    
##  [19] gower_0.2.2           pkgdown_1.6.1.9001    lpSolve_5.6.15       
##  [22] prettyunits_1.1.1     colorspace_2.0-0      blob_1.2.1           
##  [25] rappdirs_0.3.3        textshaping_0.3.3     xfun_0.22            
##  [28] dplyr_1.0.5           crayon_1.4.1          jsonlite_1.7.2       
##  [31] hexbin_1.28.2         impute_1.65.0         survival_3.2-10      
##  [34] iterators_1.0.13      glue_1.4.2            gtable_0.3.0         
##  [37] ipred_0.9-11          zlibbioc_1.37.0       XVector_0.31.1       
##  [40] kernlab_0.9-29        scales_1.1.1          vsn_3.59.1           
##  [43] mvtnorm_1.1-1         DBI_1.1.1             viridisLite_0.4.0    
##  [46] xtable_1.8-4          progress_1.2.2        clue_0.3-59          
##  [49] bit_4.0.4             proxy_0.4-25          mclust_5.4.7         
##  [52] preprocessCore_1.53.2 MsCoreUtils_1.3.3     lava_1.6.9           
##  [55] prodlim_2019.11.13    sampling_2.9          httr_1.4.2           
##  [58] FNN_1.1.3             RColorBrewer_1.1-2    ellipsis_0.3.2       
##  [61] farver_2.1.0          pkgconfig_2.0.3       nnet_7.3-15          
##  [64] sass_0.3.1            dbplyr_2.1.1          utf8_1.2.1           
##  [67] caret_6.0-86          labeling_0.4.2        tidyselect_1.1.1     
##  [70] rlang_0.4.11          reshape2_1.4.4        munsell_0.5.0        
##  [73] tools_4.1.0           LaplacesDemon_16.1.4  cachem_1.0.4         
##  [76] generics_0.1.0        RSQLite_2.2.7         evaluate_0.14        
##  [79] stringr_1.4.0         fastmap_1.1.0         mzID_1.29.0          
##  [82] yaml_2.2.1            ragg_1.1.2            ModelMetrics_1.2.2.2 
##  [85] bit64_4.0.5           fs_1.5.0              randomForest_4.6-14  
##  [88] purrr_0.3.4           KEGGREST_1.31.1       dendextend_1.14.0    
##  [91] ncdf4_1.17            nlme_3.1-152          biomaRt_2.47.8       
##  [94] compiler_4.1.0        filelock_1.0.2        curl_4.3.1           
##  [97] png_0.1-7             e1071_1.7-6           affyio_1.61.0        
## [100] tibble_3.1.1          bslib_0.2.4           stringi_1.5.3        
## [103] highr_0.9             desc_1.3.0            lattice_0.20-41      
## [106] Matrix_1.3-2          vctrs_0.3.8           pillar_1.6.0         
## [109] lifecycle_1.0.0       BiocManager_1.30.12   jquerylib_0.1.4      
## [112] MALDIquant_1.19.3     data.table_1.14.0     R6_2.5.0             
## [115] pcaMethods_1.83.0     affy_1.69.1           bookdown_0.22        
## [118] gridExtra_2.3         codetools_0.2-18      gtools_3.8.2         
## [121] MASS_7.3-53.1         assertthat_0.2.1      rprojroot_2.0.2      
## [124] withr_2.4.2           hms_1.0.0             grid_4.1.0           
## [127] rpart_4.1-15          timeDate_3043.102     coda_0.19-4          
## [130] rmarkdown_2.7         segmented_1.3-4       pROC_1.17.0.1        
## [133] lubridate_1.7.10</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor" aria-hidden="true"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Breckels2013" class="csl-entry">
Breckels, Lisa M, Laurent Gatto, Andy Christoforou, Arnoud J Groen, Kathryn S Lilley, and Matthew W B Trotter. 2013. <span>“The Effect of Organelle Discovery Upon Sub-Cellular Protein Localisation.”</span> <em>J Proteomics</em>, March. <a href="https://doi.org/10.1016/j.jprot.2013.02.019" class="external-link">https://doi.org/10.1016/j.jprot.2013.02.019</a>.
</div>
<div id="ref-Christoforou:2016" class="csl-entry">
Christoforou, A, C M Mulvey, L M Breckels, A Geladaki, T Hurrell, P C Hayward, T Naake, et al. 2016. <span>“A Draft Map of the Mouse Pluripotent Stem Cell Spatial Proteome.”</span> <em>Nat Commun</em> 7: 8992. <a href="https://doi.org/10.1038/ncomms9992" class="external-link">https://doi.org/10.1038/ncomms9992</a>.
</div>
<div id="ref-Crook:2018" class="csl-entry">
Crook, Oliver M, Claire M Mulvey, Paul D. W. Kirk, Kathryn S Lilley, and Laurent Gatto. 2018. <span>“A Bayesian Mixture Modelling Approach for Spatial Proteomics.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/282269" class="external-link">https://doi.org/10.1101/282269</a>.
</div>
<div id="ref-Dunkley2006" class="csl-entry">
Dunkley, Tom P. J., Svenja Hester, Ian P. Shadforth, John Runions, Thilo Weimar, Sally L. Hanton, Julian L. Griffin, et al. 2006. <span>“Mapping the Arabidopsis Organelle Proteome.”</span> <em>Proc Natl Acad Sci USA</em> 103 (17): 6518–23. <a href="https://doi.org/10.1073/pnas.0506958103" class="external-link">https://doi.org/10.1073/pnas.0506958103</a>.
</div>
<div id="ref-Foster2006" class="csl-entry">
Foster, Leonard J., Carmen L. de Hoog, Yanling Zhang, Yong Zhang, Xiaohui Xie, Vamsi K. Mootha, and Matthias Mann. 2006. <span>“A Mammalian Organelle Map by Protein Correlation Profiling.”</span> <em>Cell</em> 125 (1): 187–99. <a href="https://doi.org/10.1016/j.cell.2006.03.022" class="external-link">https://doi.org/10.1016/j.cell.2006.03.022</a>.
</div>
<div id="ref-Gatto:2018" class="csl-entry">
Gatto, Laurent, Lisa M Breckels, and Kathryn S Lilley. 2018. <span>“Assessing Sub-Cellular Resolution in Spatial Proteomics Experiments.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/377630" class="external-link">https://doi.org/10.1101/377630</a>.
</div>
<div id="ref-Gatto2012" class="csl-entry">
Gatto, Laurent, and Kathryn S Lilley. 2012. <span>“<span>MS</span>nbase – an <span>R</span>/<span>B</span>ioconductor Package for Isobaric Tagged Mass Spectrometry Data Visualization, Processing and Quantitation.”</span> <em>Bioinformatics</em> 28 (2): 288–89. <a href="https://doi.org/10.1093/bioinformatics/btr645" class="external-link">https://doi.org/10.1093/bioinformatics/btr645</a>.
</div>
<div id="ref-Gatto2010" class="csl-entry">
Gatto, Laurent, Juan Antonio Vizcaíno, Henning Hermjakob, Wolfgang Huber, and Kathryn S Lilley. 2010. <span>“Organelle Proteomics Experimental Designs and Analysis.”</span> <em>Proteomics</em>. <a href="https://doi.org/10.1002/pmic.201000244" class="external-link">https://doi.org/10.1002/pmic.201000244</a>.
</div>
<div id="ref-Gentleman2004" class="csl-entry">
Gentleman, Robert C., Vincent J. Carey, Douglas M. Bates, Ben Bolstad, Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004. <span>“Bioconductor: Open Software Development for Computational Biology and Bioinformatics.”</span> <em>Genome Biol</em> 5 (10): –80. <a href="https://doi.org/10.1186/gb-2004-5-10-r80" class="external-link">https://doi.org/10.1186/gb-2004-5-10-r80</a>.
</div>
<div id="ref-tsne:2008" class="csl-entry">
Maaten, Laurens van der, and Geoffrey Hinton. 2008. <span>“<span class="nocase">Visualizing Data using t-SNE</span>.”</span> <em>Journal of Machine Learning Research</em> 9 (November): 2579–2605. <a href="http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf" class="external-link">http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf</a>.
</div>
<div id="ref-Rstat" class="csl-entry">
R Development Core Team. 2011. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria: R Foundation for Statistical Computing. <a href="http://www.R-project.org/" class="external-link">http://www.R-project.org/</a>.
</div>
<div id="ref-Ross2004" class="csl-entry">
Ross, Philip L., Yulin N. Huang, Jason N. Marchese, Brian Williamson, Kenneth Parker, Stephen Hattan, Nikita Khainovski, et al. 2004. <span>“Multiplexed Protein Quantitation in Saccharomyces Cerevisiae Using Amine-Reactive Isobaric Tagging Reagents.”</span> <em>Mol Cell Proteomics</em> 3 (12): 1154–69. <a href="https://doi.org/10.1074/mcp.M400129-MCP200" class="external-link">https://doi.org/10.1074/mcp.M400129-MCP200</a>.
</div>
<div id="ref-shiny" class="csl-entry">
RStudio, and Inc. 2014. <em>Shiny: Web Application Framework for r</em>. <a href="http://CRAN.R-project.org/package=shiny" class="external-link">http://CRAN.R-project.org/package=shiny</a>.
</div>
<div id="ref-Tan2009" class="csl-entry">
Tan, Denise J, Heidi Dvinge, Andy Christoforou, Paul Bertone, Alfonso A Martinez, and Kathryn S Lilley. 2009. <span>“Mapping Organelle Proteins and Protein Complexes in Drosophila Melanogaster.”</span> <em>J Proteome Res</em> 8 (6): 2667–78. <a href="https://doi.org/10.1021/pr800866n" class="external-link">https://doi.org/10.1021/pr800866n</a>.
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>The content of this document is compiled (the code is executed and its output, text and figures, is displayed dynamically) to generate the pdf file.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Briefly, the <code>itraqdata</code> raw iTRAQ4-plex data is quantified by trapezoidation using <em><a href="https://bioconductor.org/packages/3.13/MSnbase" class="external-link">MSnbase</a></em> functionality. See the <code>MSnbase-demo</code> vignette for details.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p><a href="https://lvdmaaten.github.io/tsne/" class="external-link uri">https://lvdmaaten.github.io/tsne/</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p><a href="http://distill.pub/2016/misread-tsne/" class="external-link uri">http://distill.pub/2016/misread-tsne/</a><a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Laurent Gatto, Lisa Breckels, Oliver Crook.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
